crucible/src/Lang/Crucible/Backend/Prove.hs

```diff
@@ -3,147 +3,452 @@ Module      : Lang.Crucible.Backend.Prove
 
 import qualified What4.Interface as W4
 import qualified What4.Expr as WE
-import qualified What4.Solver.Adapter as WSA
+import qualified What4.Protocol.Online as WPO
+import qualified What4.Protocol.SMTWriter as W4SMT
 import qualified What4.SatResult as W4R
+import qualified What4.Solver.Adapter as WSA
 
 import qualified Lang.Crucible.Backend as CB
 import           Lang.Crucible.Backend.Assumptions (Assumptions)
-import           Lang.Crucible.Backend.ProofGoals (traverseGoalsWithAssumptions)
+import           Lang.Crucible.Utils.Timeout (Timeout, TimedOut)
+import qualified Lang.Crucible.Utils.Timeout as CTO
+
+-- | Local helper
+consumeGoals ::
+  -- | Consume an 'Assuming'
+  (asmp -> a -> a) ->
+  -- | Consume a 'Prove'
+  (goal -> a) ->
+  -- | Consume a 'ProveConj'
+  (a -> a -> a) ->
+  CB.Goals asmp goal ->
+  a
+consumeGoals onAssumption onGoal onConj = go
+  where
+  go (CB.Prove gl) = onGoal gl
+  go (CB.Assuming as gl) = onAssumption as (go gl)
+  go (CB.ProveConj g1 g2) = onConj (go g1) (go g2)
+
+-- | Local helper
+consumeGoalsWithAssumptions ::
+  forall asmp goal a.
+  Monoid asmp =>
+  -- | Consume a 'Prove'
+  (asmp -> goal -> a) ->
+  -- | Consume a 'ProveConj'
+  (a -> a -> a) ->
+  CB.Goals asmp goal ->
+  a
+consumeGoalsWithAssumptions onGoal onConj goals =
+  Reader.runReader (go goals) mempty
+  where
+  go :: CB.Goals asmp goal -> Reader.Reader asmp a
+  go =
+    consumeGoals
+      (\asmp gl -> Reader.local (<> asmp) gl)
+      (\gl -> Reader.asks (\asmps -> onGoal asmps gl))
+      (\g1 g2 -> onConj <$> g1 <*> g2)
+
+---------------------------------------------------------------------
+-- * Strategy
 
 -- | The result of attempting to prove a goal with an SMT solver.
 --
 -- The constructors of this type correspond to those of 'W4R.SatResult'.
+--
+-- * @sym@ is the symbolic backend, usually 'WE.ExprBuilder'
+-- * @r@ is the return type of the eventual 'ProofConsumer'
 data ProofResult sym t
    = -- | The goal was proved.
      --
      -- Corresponds to 'W4R.Unsat'.
      Proved
      -- | The goal was disproved, and a model that falsifies it is available.
      --
+     -- The 'WE.GroundEvalFn' is only available for use during the execution of
+     -- a 'ProofConsumer'. See 'WSA.SolverAdapter'.
+     --
+     -- The @'Maybe' 'WE.ExprRangeBindings'@ are 'Just' when using
+     -- 'offlineProve' and 'Nothing' when using 'onlineProve'.
+     --
      -- Corresponds to 'W4R.Sat'.
    | Disproved (WE.GroundEvalFn t) (Maybe (WE.ExprRangeBindings t))
      -- | The SMT solver returned \"unknown\".
      --
      -- Corresponds to 'W4R.Unknown'.
    | Unknown
 
--- | Prove a single goal ('CB.Assertion') under the supplied 'Assumptions'.
+-- | A 'ProofStrategy' dictates how results are proved.
 --
--- The overall approach is:
+-- * @sym@ is the symbolic backend, usually 'WE.ExprBuilder'
+-- * @m@ is the monad in which the 'Prover' and 'Combiner' run
+-- * @t@ is the parameter to 'WE.Expr'
```

Comment:
It might be worth clarifying that `t` is a What4 `BaseType`.

Reply:
It actually [is *not* a `BaseType`](https://hackage.haskell.org/package/what4-1.6/docs/What4-Expr.html#t:Expr), which is perhaps itself worth clarifying.